components/brave_ads/core/internal/user_engagement/ad_events/search_result_ads/search_result_ad_event_handler_util.cc

```diff
@@ -0,0 +1,100 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/internal/user_engagement/ad_events/search_result_ads/search_result_ad_event_handler_util.h"
+
+#include "base/check.h"
+#include "brave/components/brave_ads/core/internal/ad_units/search_result_ad/search_result_ad_info.h"
+#include "brave/components/brave_ads/core/internal/common/logging_util.h"
+#include "brave/components/brave_ads/core/internal/creatives/conversions/creative_set_conversion_builder.h"
+#include "brave/components/brave_ads/core/internal/creatives/conversions/creative_set_conversion_database_table_util.h"
+#include "brave/components/brave_ads/core/internal/creatives/conversions/creative_set_conversion_info.h"
+#include "brave/components/brave_ads/core/internal/settings/settings.h"
+#include "brave/components/brave_ads/core/internal/user_engagement/ad_events/ad_event_handler_util.h"
+#include "brave/components/brave_ads/core/mojom/brave_ads.mojom.h"  // IWYU pragma: keep
+#include "brave/components/brave_ads/core/public/ads_feature.h"
+
+namespace brave_ads {
+
+void MaybeBuildAndSaveCreativeSetConversion(
+    const mojom::CreativeSearchResultAdInfoPtr& mojom_creative_ad,
+    const mojom::SearchResultAdEventType event_type) {
+  const bool user_has_joined_rewards = UserHasJoinedBraveRewards();
+
+  // Save only if:
+  // - User has joined Brave Rewards and a viewed impression event was fired.
+  // - User has not joined Brave Rewards and a clicked impression event was
+  //   fired.
+
+  const bool fired_view_event_for_rewards_user =
+      user_has_joined_rewards &&
+      event_type == mojom::SearchResultAdEventType::kViewedImpression;
+
+  const bool fired_click_event_for_non_rewards_user =
+      !user_has_joined_rewards &&
```

Comment:
For non-Rewards users, we should only save the creative set conversions for clicks, whereas for Rewards users, we should only save the creative set conversions for views. This is because the creative set conversion is identical for both events.